### DISCUSSION TITLE
Pattern to trigger one callback per page when some block is present

### DISCUSSION THREAD

**Discussion by luisherranz** [WordPress/gutenberg#52514] [Interactivity API]
Sometimes, blocks need to do something global on a page.

Imagine a block that needs to run some util from an external library. If only one block is included in the page, we could get away with `data-wp-init`:

```html
<div class="wp-block-my-block" data-wp-init="init.namespace.runLibrary">
  ...
</div>
```

```js
import { library } from "external-library";

store({
  init: {
    namespace: {
      runLibrary: ({ state }) => {
        library(state.namespace.someValue);
      },
    },
  },
});
```

But if multiple blocks are included, the `data-wp-init` will be run multiple times.

We could add some initialization flag, like this:

```js
let init = false;

store({
  init: {
    namespace: {
      runLibrary: ({ state }) => {
        if (!init) {
          library(state.namespace.someValue);
          init = true;
        }
      },
    },
  },
});
```

But it's not very clean and we will include `data-wp-init` in the HTML more times than necessary.

One idea that we have is to let store initialization be a function:

```js
store(({ state }) => {
  // Init.
  library(state.namespace.someValue);

  return {
    actions: {
      /* ... */
    },
    /* ... */
  };
});
```

It's not bad, although in this case, `state` may not be fully initialized if `someValue` comes from another `store()` call, so people would need to be careful about that, and it might force us into introducing some sort of priority system for the `store()` calls.

Also, there may be a case where this function needs to be run, not only by one block but by many blocks. In that case, I guess all the blocks should import the same JavaScript file that does a call to `store()` to initialize the external library.

Another idea could be to let blocks add a `data-wp-init` directive to the `body`, for example:

```html
<body data-wp-init--run-library="init.namespace.runLibrary">
  ...
</body>
```

With a PHP API like this:

```php
wp_add_body_directive( 'data-wp-init--run-library', 'init.namespace.runLibrary' );
```

Any other ideas?

**Comment by DAreRodz** [comment]
Back in the "old days", for that kind of initialization, we had in Frontity some [lifecycle initialization actions](https://docs.frontity.org/learning-frontity/actions#frontity-lifecycle-initialization-actions): `init`, `beforeCSR`, `afterCSR`, etc.

They helped packages do their stuff in different moments of the Frontity set-up. And it is pretty much what we want to do here, am I right? 😄

So, we could use that approach here as well, e.g.:

```js
store({
  actions: {
    namespace: {
      // This special action could run after the store is ready.
      init: ({ state }) => {
        library(state.namespace.someValue);
      },
    },
  },
});
```


Another thing I have in mind (and I prefer) is to add a second argument to the `store` function, where you can define these kinds of callbacks, e.g.,

```js
store(
  {
    state: {
      namespace: { /* &c. */ },
    },
  },
  // Optional object with callback definitions.
  {
    init: ({ state }) => {
      library(state.namespace.someValue);
    },
  }
);
```

Those would receive the whole store as argument and run once at the appropriate moment.

For the case you mentioned, @luisherranz, in which different types of blocks could want to run the same _initialization_, you could solve that by defining an action and calling it inside the `init` callback for each block type.

```js
// Block A
store({
  actions: {
    namespace: {
      init: ({ state }) => {
        library(state.namespace.someValue);
      },
    },
  },
});

// Block B, C, etc.
store({ /* store props (if any) */ }, {
  init: (store) {
    store.actions.namespace.init(store);
  },
});
```

We could also define different initialization callbacks (`init`, `hydration`, _`whatever`_) for different lifecycle hooks.

In any case, for the initialization purposes you mentioned, I don't think requiring adding a directive in the HTML would be a good DX. If it is for certain block instances, though, then it's OK. 👍

**Reply by luisherranz** [reply]
I like this idea!

[I've been thinking about how lazy loading would work](https://github.com/WordPress/gutenberg/discussions/52723) because I think they should work together: i.e., this callback should not be executed until the block is hydrated. I think that this is solved if we delay the execution of the store until the block is about to be hydrated, as I explained in the discussion.

I also like the idea of moving this callback to the `options` (second parameter) of the `store` function.

Regarding the naming, as [I said in the lazy loading discussion](https://github.com/WordPress/gutenberg/discussions/52723), I suggest using `"load"` to refer to the "hydration" of a block, so I think we could use `"beforeLoad"` and `"afterLoad"` here.

```js
store(
  {
    // regular store..
  },
  {
    beforeLoad: async ({ state }) => {
      // Something that needs to happen before the hydration.
    },
    afterLoad: async ({ state }) => {
      // Something that needs to happen after the hydration.
    },
  }
);
```

`beforeLoad` should be used with care as it will block the hydration. If the block is lazy-loaded, it's not that important as it won't affect other blocks. And depending on how we implement [the lazy loading](https://github.com/WordPress/gutenberg/discussions/52723), we may do it so we reuse that logic and `beforeLoad` only blocks the hydration of a particular block even on initial load. But if the site has full-page client-side navigation, all the blocks need to be hydrated at the beginning, meaning that any `beforeLoad` will block the hydration of all the blocks of the site.

For that reason, I think I would start with `afterLoad` and wait until we see real use cases for `beforeLoad`.

**Reply by luisherranz** [reply]
Oh, we may also need an `afterEachLoad` callback if they need to do something specific to each DOM reference:

```js
store(
  {
    // regular store..
  },
  {
    afterEachLoad: async ({ state, ref }) => {
      // Something that needs to happen after the hydration, once per block and
      // with access to the DOM reference.
    },
  }
);
```

But again, let's wait until we see real use cases for that one.

**Reply by DAreRodz** [reply]
> Oh, we may also need an `afterEachLoad` callback if they need to do something specific to each DOM reference

I think this is very similar to `data-wp-init`, isn't it? What would be the differences between them? 🤔

**Reply by luisherranz** [reply]
Hmm... I guess you're right 🙂

**Comment by SantosGuillamot** [comment]
I feel the approach David suggests seems easier regarding DX 🙂 Does it make sense to include the different initialization callbacks (`init` / `hydration` / `whatever`) in the store, or should they be another function? I'm not fully familiarized with these patterns, so you will know better than me.

On another note, maybe calling everything `init`, added to the `wp-init` directive, might be confusing for developers?

**Reply by DAreRodz** [reply]
> On another note, maybe calling everything `init`, added to the `wp-init` directive, might be confusing for developers?

Do you mean using a different name rather than `init` for the initialization callback? Do you have anything in mind?

**Reply by SantosGuillamot** [reply]
Or just clarifying the differences. I didn't think too much about it. But I can imagine people wondering the difference about using a `wp-init` directive or using init there.

**Reply by luisherranz** [reply]
Let me know what you think about `beforeLoad` and `afterLoad`: [link](https://github.com/WordPress/gutenberg/discussions/52514#discussioncomment-6476546)

**Reply by SantosGuillamot** [reply]
I like `beforeLoad` and `afterLoad` 🙂 I feel it will be easy to understand their purpose.

**Comment by luisherranz** [comment]
> `wp_add_body_directive( 'data-wp-init--run-library', 'init.namespace.runLibrary' );`

I think we should discard the `wp_add_body_directive` idea for now because people could start using this API to add context to the whole page, which would be an anti-pattern:

```php
wp_add_body_directive( 'data-wp-context', '{ ... some global state ... }' );
```

**Comment by luisherranz** [comment]
@darerodz started a PR to add this functionality:

- https://github.com/WordPress/gutenberg/pull/53363

**Comment by luisherranz** [comment]
This is no longer needed with the new `store()` API because the `state` is accessible from the beginning. It's true that blocks might still need to wait until all the other store parts are loaded, but maybe they can do so with a `DOMContentLoaded` or a `requestIdleCallback` call.

- https://github.com/WordPress/gutenberg/discussions/53586

We will remove it from the API.
